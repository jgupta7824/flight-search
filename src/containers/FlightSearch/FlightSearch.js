import React, { Fragment, useReducer, useState } from 'react'
import classes from './FlightSearch.module.css'
import FlightResult from '../FlightSearchResult/FlightResults'
import TRANSLATIONS from '../../utils/translations'
import SelectBox from '../../components/SelectBox.component'
import Date from '../../components/Date.component'
import { DepartFormData, GoingToFormData, classData } from '../../utils/mockData'
import { enableSeachResults, updateSearchFormData } from '../../actions/AppActions'
import { useDispatch, useSelector } from 'react-redux'
import { getTodaysDate } from '../../utils/CommonFile'

let intialState = {
  departFrom: 'Chandigarh (IXC)',
  goingTo: 'Mumbai (BOM)',
  class: 'Economy',
  travellers: 1,
  fromDate: getTodaysDate(),
  toDate: getTodaysDate()
}

function reducer(state, action) {
  switch (action.type) {
    case 'updateDepartFrom':
      return { ...state, departFrom: action.value };
    case 'updateGoingTo':
      return { ...state, goingTo: action.value };
    case 'updateClass':
      return { ...state, class: action.value };
    case 'updateTravellers':
      return { ...state, travellers: action.value };
    case 'updateFromDate':
      return { ...state, fromDate: action.value };
    case 'updateToDate':
      return { ...state, toDate: action.value };
    default:
      throw new Error();
  }
}

const FlightSearch = (props) => {

  const isSearchEnabled = useSelector(state => state.isSearchEnabled)
  const formData = useSelector(state => state.formData)
  const [state, updateState] = useReducer(reducer, intialState);
  const [dateError, setDateError] = useState('')
  const dispatch = useDispatch();

  //Search Flight
  function searchFlights(event) {
    event.preventDefault()
    if (!state.fromDate && !state.toDate) {
      setDateError(TRANSLATIONS.PROVIDE_FROMDATE_TODATE)
      return
    } else if (!state.fromDate && state.toDate) {
      setDateError(TRANSLATIONS.FROMDATE_CAN_NOT_BE_EMPTY)
      return
    } else if (window.Date.parse(state.fromDate) > window.Date.parse(state.toDate)) {
      setDateError(TRANSLATIONS.FROMDATE_CAN_NOT_BE_GREATER)
      return
    }
    setDateError('')
    console.log(state)
    dispatch(updateSearchFormData(state))
    dispatch(enableSeachResults(true))
  }

  //update state on Change departTo dropdown 
  //can be used refs but not using since management of refs is not good

  function onChangeSelect(event, type) {
    if (event.target.value) {
      updateState({ type, value: event.target.value || '' })
    }
  }

  return (
    <Fragment>
      {!isSearchEnabled && <div className={classes.panel} id="container">
        <div className={classes.formContainer}>
          <form>
            <div className={classes.formGroup}>
              <label className={classes.formLabel}>{TRANSLATIONS.DEPART_FORM}</label>
              <SelectBox data={DepartFormData}
                defaultValue={formData && formData['departFrom'] ? formData['departFrom'] : state.departFrom}
                onChange={(event) => onChangeSelect(event, 'updateDepartFrom')} />
            </div>
            <div className={classes.formGroup}>
              <label className={classes.formLabel}>{TRANSLATIONS.GOING_TO}</label>
              <SelectBox data={GoingToFormData}
                defaultValue={formData && formData['goingTo'] ? formData['goingTo'] : state.goingTo}
                onChange={(event) => onChangeSelect(event, 'updateGoingTo')} />
            </div>
            <div className="rowFlex">
              <div>
                <label className={classes.formLabel}>{TRANSLATIONS.DEPART_DATE}</label>
                <Date style={{ width: '175px', marginRight: '20px' }}
                  min={getTodaysDate()}
                  value={formData && formData['fromDate'] ? formData['fromDate'] : getTodaysDate()}
                  onChange={(event) => onChangeSelect(event, 'updateFromDate')} />
              </div>
              <div >
                <label className={classes.formLabel}>{TRANSLATIONS.RETURN_DATE}</label>
                <Date style={{ width: '175px' }} onChange=
                  {(event) => onChangeSelect(event, 'updateToDate')}
                  value={formData && formData['toDate'] ? formData['toDate'] : getTodaysDate()}
                  min={getTodaysDate()}
                />
              </div>
            </div>
            <div className="rowFlex">
              <div>
                <label className={classes.formLabel}>{TRANSLATIONS.TRAVELLERS}</label>
                <input type="number" id="travellers"
                  value={formData && formData['travellers'] ? formData['travellers'] : state.travellers}
                  name="travellers" min="1" max="100" className={classes.inputStyle}
                  onChange={(event) => onChangeSelect(event, 'updateTravellers')}
                  style={{ marginRight: '20px', width: '170px' }} />
              </div>
              <div >
                <label className={classes.formLabel}>{TRANSLATIONS.CLASS}</label>
                <SelectBox style={{ width: '190px' }} data={classData}
                  defaultValue={formData && formData['class'] ? formData['class'] : state.class}
                  onChange={(event) => onChangeSelect(event, 'updateClass')} />
              </div>
            </div>
            {dateError && <p style={{ color: 'red' }}>{dateError}</p>}
            <div className={classes.buttonContainer}>
              <button className="buttonStyle" onClick={searchFlights}>Search Flights</button>
            </div>
          </form>
        </div>
      </div>}
      {isSearchEnabled && <FlightResult />}
    </Fragment>
  )
}

export default FlightSearch;