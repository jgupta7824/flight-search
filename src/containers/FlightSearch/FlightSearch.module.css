.panel {
    margin: 105px 25px 25px;
    width:480px;
    -webkit-box-shadow: 0 0 25px rgb(0 0 0 / 30%);
	  -moz-box-shadow: 0 0 25px rgb(0 0 0 / 30%);
     box-shadow: 0 0 25px rgb(0 0 0 / 30%);
     position: absolute;
     left: 50%;
     transform: translateX(-50%) translateY(0%);
  }

  .formContainer{
    display: flex;
    flex-direction: column;
    width: 100%;
    height: 100%;
    padding: 10px 10px 10px 30px;
  }

  .formGroup{
    display: flex;
    flex-direction: column;
    width:400px
  }

  .formLabel{
    font-weight: 600;
    font-size: .9rem;
    display: block;
    margin: .5rem 0;
  }

  .inputStyle{
    padding:8px;
  }

  .dateInput{
    width: 170px;
    padding:5px;
  }

  .buttonContainer{
    padding: 20px;
    margin-right: 95px;
    text-align: center;
  }

  [type="date"] {
    background:
    #fff url(https://cdn1.iconfinder.com/data/icons/cc_mono_icon_set/blacks/16x16/calendar_2.png) 
     97% 50% no-repeat ;
  }
  [type="date"]::-webkit-inner-spin-button {
    display: none;
  }
  [type="date"]::-webkit-calendar-picker-indicator {
    opacity: 0;
  }