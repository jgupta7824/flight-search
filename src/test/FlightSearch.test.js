import React from "react";
import FlightSearch from "../containers/FlightSearch/FlightSearch";
import { configure, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'


configure({ adapter: new Adapter() })
jest.mock('react-redux', () => ({
    useSelector: jest.fn(fn => fn()),
    useDispatch: () => jest.fn(),
    connect: () => jest.fn()
}))

describe("Render FlightSearch Component", () => {
    it("render FlightSearch  page", () => {
        const FlightSearchComponent = shallow(<FlightSearch />);
        expect(toJson(FlightSearchComponent)).toMatchSnapshot()
    });

});