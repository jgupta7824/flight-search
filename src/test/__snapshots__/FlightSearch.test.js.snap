// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render FlightSearch Component render FlightSearch  page 1`] = `
<Fragment>
  <div
    className="panel"
    id="container"
  >
    <div
      className="formContainer"
    >
      <form>
        <div
          className="formGroup"
        >
          <label
            className="formLabel"
          >
            Depart Form
          </label>
          <SelectBox
            data={
              Array [
                Object {
                  "id": "IXC",
                  "name": "Chandigarh (IXC)",
                },
                Object {
                  "id": "BLR",
                  "name": "Bangalore (BLR)",
                },
                Object {
                  "id": "MAA",
                  "name": "Chennai (MAA)",
                },
                Object {
                  "id": "DEL",
                  "name": "New Delhi (DEL)",
                },
                Object {
                  "id": "JAI",
                  "name": "Jaipur (JAI)",
                },
              ]
            }
            defaultValue="Chandigarh (IXC)"
            onChange={[Function]}
          />
        </div>
        <div
          className="formGroup"
        >
          <label
            className="formLabel"
          >
            Going To
          </label>
          <SelectBox
            data={
              Array [
                Object {
                  "id": "BOM",
                  "name": "Mumbai (BOM)",
                },
                Object {
                  "id": "GOI",
                  "name": "Goa (GOI)",
                },
                Object {
                  "id": "MAA",
                  "name": "Kolkata (CCU)",
                },
                Object {
                  "id": "PNQ",
                  "name": "Pune (PNQ)",
                },
                Object {
                  "id": "KNU",
                  "name": "Kanpur (KNU)",
                },
              ]
            }
            defaultValue="Mumbai (BOM)"
            onChange={[Function]}
          />
        </div>
        <div
          className="rowFlex"
        >
          <div>
            <label
              className="formLabel"
            >
              Depart Date
            </label>
            <Date
              min="2021-08-20"
              onChange={[Function]}
              style={
                Object {
                  "marginRight": "20px",
                  "width": "175px",
                }
              }
              value="2021-08-20"
            />
          </div>
          <div>
            <label
              className="formLabel"
            >
              Return Date
            </label>
            <Date
              min="2021-08-20"
              onChange={[Function]}
              style={
                Object {
                  "width": "175px",
                }
              }
              value="2021-08-20"
            />
          </div>
        </div>
        <div
          className="rowFlex"
        >
          <div>
            <label
              className="formLabel"
            >
              Travelers
            </label>
            <input
              className="inputStyle"
              id="travellers"
              max="100"
              min="1"
              name="travellers"
              onChange={[Function]}
              style={
                Object {
                  "marginRight": "20px",
                  "width": "170px",
                }
              }
              type="number"
              value={1}
            />
          </div>
          <div>
            <label
              className="formLabel"
            >
              Class
            </label>
            <SelectBox
              data={
                Array [
                  Object {
                    "id": "economy",
                    "name": "Economy",
                  },
                  Object {
                    "id": "basic",
                    "name": "Basic",
                  },
                  Object {
                    "id": "main",
                    "name": "Main",
                  },
                ]
              }
              defaultValue="Economy"
              onChange={[Function]}
              style={
                Object {
                  "width": "190px",
                }
              }
            />
          </div>
        </div>
        <div
          className="buttonContainer"
        >
          <button
            className="buttonStyle"
            onClick={[Function]}
          >
            Search Flight
          </button>
        </div>
      </form>
    </div>
  </div>
</Fragment>
`;
